scene/animation/animation_node_state_machine.cpp

```diff
@@ -1259,13 +1259,20 @@ bool AnimationNodeStateMachine::is_parameter_read_only(const StringName &p_param
 }
 
 void AnimationNodeStateMachine::add_node(const StringName &p_name, Ref<AnimationNode> p_node, const Vector2 &p_position) {
+	//void AnimationNodeStateMachine::add_node(const StringName &p_name, Ref<AnimationNode> p_node, const Vector2 &p_position, const PackedInt32Array &p_layer_options, const bool &p_node_visibility) {
 	ERR_FAIL_COND(states.has(p_name));
 	ERR_FAIL_COND(p_node.is_null());
 	ERR_FAIL_COND(String(p_name).contains("/"));
 
+	PackedInt32Array layer_options;
+	layer_options.resize(12);
+	layer_options.fill(0);
+
 	State state_new;
 	state_new.node = p_node;
 	state_new.position = p_position;
+	state_new.layer_options = layer_options;
+	state_new.node_visibility = true;
```

Comment:
```suggestion
	State state_new;
	state_new.node = p_node;
	state_new.position = p_position;
#ifdef TOOLS_ENABLED
	state_new.layer_options = layer_options;
	state_new.node_visibility = true;

	PackedInt32Array layer_options;
	layer_options.resize(12);
	layer_options.fill(0);
#endif
```
They should all be done in TOOLS_ENABLED block unless they change the runtime behavior. See also the behavior of Handle in BezierTrack.

Reply:
Fair point. That's not what I implemented, so it's just a guess, but position probably exposes it because when you add an animation node externally, it can get messed up when opened in an editor. We may be able to rise an issue/discussion to make it TOOLS_ENABLED only. Nevertheless, I think it should be TOOLS_ENABLED because it is difficult to manage the tags externally.

Reply:
I have added TOOLS_ENABLED to pretty much everything except for setters and getters just incase if anyone wanna get them for grouping animations and get group names for debugging or organizing.

Reply:
As you can see from the other Nodes, it is also required for TOOLS_ENABLED setters. Instead, we need to indicate in the documentation that it is an editor-only option.

Reply:
So anything that doesn't need to be accessed or get during game play is TOOLS_ENABLED. Ex; you won't add new animations during game is running (runtime) or add animations into new groups? Even letting getters go for animation groups is kinda generous and maybe a bit just in case.. 
Which would also include pretty much most of the things in editor menus? So to clarify I will just put everything I have added under TOOLS_ENABLED except for getters and things that interacts with other items  to avoid making further changes (which is just a few line for editor menu).
Other than these is there anything I am currently missing or getting wrong for this change request?

Reply:
The main problem is that the layer is changing the behavior of the StateMachine, since changing the behavior of a StateMachine should essentially be implemented through the StateMachineEditor, and TOOLS_ENABLED is just a hack.

Reply:
But the functional/behavioral changes are made in `animation_state_machine_editor`; hides and shows nodes and organize groups.
`animation_node_state_machine` is just used for storing and getting properties, and since it works with the save folder for state machine and has access to individual nodes I thought I should be using this for saving changes and getting individual nodes.

Reply:
> But the functional/behavioral changes are made in animation_state_machine_editor; hides and shows nodes and organize groups.

Okay looking at the whole thing, it looks fine thet animation_node_state_machine.cpp does not control the playback. But the data still needs to be TOOLS_ENABLED.

Reply:
> But the functional/behavioral changes are made in animation_state_machine_editor; hides and shows nodes and organize groups.

Okay looking at the whole thing, it looks fine thet animation_node_state_machine.cpp does not control the playback. But the data still needs to be TOOLS_ENABLED.